storage/range.go

```diff
@@ -1548,3 +1593,109 @@ func (r *Range) SetHardState(st raftpb.HardState) error {
 	return engine.MVCCPutProto(r.rm.Engine(), nil, engine.RaftStateKey(r.Desc.RaftID),
 		proto.ZeroTimestamp, nil, &st)
 }
+
+// intersectReplicaSets is used in AdminMerge to ensure that the ranges are
+// all colocated on the same set of replicas.
+func intersectReplicaSets(a, b []proto.Replica) bool {
+	if len(a) != len(b) {
+		return false
+	}
+
+	set := make(map[int32]int)
+	for _, replica := range a {
+		set[replica.StoreID]++
+	}
+
+	for _, replica := range b {
+		set[replica.StoreID]--
+	}
+
+	for _, value := range set {
+		if value != 0 {
+			return false
+		}
+	}
+
+	return true
+}
+
+// AdminMerge extends the range to subsume the range that comes next in
+// the key space.  The range being subsumed is provided in args.SubsumedRange.
+// The EndKey of the the subsuming range must equal the start key of the
+// range being subsumed.  The merge is performed inside of a distributed
+// transaction which writes the updated range descriptor for the subsuming range
+// and deletes the range descriptor for the subsumed one.  It also updates the
+// range addressing metadata.  The handover of responsibility for
+// the reassigned key range is carried out seamlessly through a merge trigger
+// carried out as part of the commit of that transaction.
+// A merge requires that the two ranges are colocated on the same set of replicas.
+func (r *Range) AdminMerge(args *proto.AdminMergeRequest, reply *proto.AdminMergeResponse) {
+	// Only allow a single split/merge per range at a time.
+	if !atomic.CompareAndSwapInt32(&r.splitting, int32(0), int32(1)) {
+		reply.SetGoError(util.Errorf("already splitting/merging range %d", r.Desc.RaftID))
+		return
+	}
+	defer func() { atomic.StoreInt32(&r.splitting, int32(0)) }()
+
+	subsumedDesc := args.SubsumedRange
+
+	// Ensure that the subsumedRange is not the first range.
+	if subsumedDesc.StartKey.Equal(engine.KeyMin) {
+		reply.SetGoError(util.Error("Cannot subsume the first range"))
+		return
+	}
+
+	// Make sure the range being subsumed follows this one.
+	if !bytes.Equal(r.Desc.EndKey, subsumedDesc.StartKey) {
+		reply.SetGoError(util.Errorf("Ranges that are not adjacent cannot be merged, %d = %d",
+			r.Desc.EndKey, subsumedDesc.StartKey))
+		return
+	}
+
+	// Ensure that both ranges are colocated by intersecting the store ids from
```

Comment:
Probably better to instead duplicate the check in the MergeTrigger method. That would have the same effect but not require a new command type (and also avoid introducing the problem of always ensuring BeginMerge is _always_ followed with the merge).